### PR TITLE
Bug fix to get correct x values in SaveNexus when spec indices

### LINKED ISSUE #33152: Issue with SaveNExus/SaveNExusProcessed with workspace indices supplied
**Describe the bug**
The x values of saved spectra with ragged x-values seem to be wrong - found while investigating #33129

**To Reproduce**
```
# import mantid algorithms, numpy and matplotlib
from mantid.simpleapi import *
import matplotlib.pyplot as plt
import numpy as np
from mantid.api import AnalysisDataService as ADS
from mantid.plots.utility import MantidAxType

CreateSampleWorkspace(OutputWorkspace='ws')
ConvertUnits(InputWorkspace='ws', OutputWorkspace='ws', Target='dSpacing')  # so peaks don't line up
# Save and load spectra seperately
SaveNexus(InputWorkspace='ws', Filename='C:/Users/xhg73778/Desktop/ws_1.nxs', 
    WorkspaceIndexList='1')
LoadNexus(Filename='C:/Users/xhg73778/Desktop/ws_1.nxs', OutputWorkspace='ws_1')
SaveNexus(InputWorkspace='ws', Filename='C:/Users/xhg73778/Desktop/ws_2.nxs', 
    WorkspaceIndexList='2')
LoadNexus(Filename='C:/Users/xhg73778/Desktop/ws_2.nxs', OutputWorkspace='ws_2')
# Save and Load all
SaveNexus(InputWorkspace='ws', Filename='C:/Users/xhg73778/Desktop/ws_all.nxs')
LoadNexus(Filename='C:/Users/xhg73778/Desktop/ws_1_2.nxs', OutputWorkspace='ws_all')

# make plot
ws_all = ADS.retrieve('ws_all')
ws_1 = ADS.retrieve('ws_1')
ws_2 = ADS.retrieve('ws_2')
ws = ADS.retrieve('ws')

fig, axes = plt.subplots(edgecolor='#ffffff', num='ws-1', subplot_kw={'projection': 'mantid'})
axes.plot(ws, color='#1f77b4', distribution=True, label='ws: spec 2', wkspIndex=1)
axes.plot(ws, color='#ff7f0e', distribution=True, label='ws: spec 3', wkspIndex=2)
axes.plot(ws_2, color='#9467bd', distribution=True, label='ws_2: spec 3', wkspIndex=0)
axes.plot(ws_all, color='#d62728', distribution=True, label='ws_all: spec 3', linestyle='--', wkspIndex=2, zorder=2.1)
axes.plot(ws_all, color='#2ca02c', distribution=True, label='ws_all: spec 2', linestyle='--', wkspIndex=1, zorder=2.1)
axes.plot(ws_1, color='#8c564b', distribution=True, label='ws_1: spec 2', wkspIndex=0, zorder=2.1)
axes.tick_params(axis='x', which='major', **{'gridOn': False, 'tick1On': True, 'tick2On': False, 'label1On': True, 'label2On': False, 'size': 6, 'tickdir': 'out', 'width': 1})
axes.tick_params(axis='y', which='major', **{'gridOn': False, 'tick1On': True, 'tick2On': False, 'label1On': True, 'label2On': False, 'size': 6, 'tickdir': 'out', 'width': 1})
axes.set_title('ws')
axes.set_xlabel('d-Spacing ($\\AA$)')
axes.set_ylabel('Counts')
legend = axes.legend(fontsize=8.0).set_draggable(True).legend
fig.show()
```

**Expected behavior**
Spectra loaded would match spectra saved

**Screenshots**
image

**Platform/Version (please complete the following information):**
 - OS: [e.g. Windows, RHEL 7, Ubuntu, macOS]
 - OS Version:
 - Mantid Version [e.g. 6.0.0]

**Additional context**

### PR DESCRIPTION
**Pull request by RichardWaiteSTFC**
**Description of work.**
Thanks @DannyHindson for this one-line fix.

This corrects the spectrum index from which the x-values of a spectrum are retrieved. This applies to SaveNexus-like algorithms called on workspaces that have ragged x bins and for which the user supplies specific spectrum indices to save. It was looping over a the spectrum indices (either supplied directly or via a min/max) but saving the x-values of the counter rather than the spectrum indices themselves.

**To test:**
Follow instructions in original issue #33152 - check loaded spectra match up with the original workspace.

Fixes #33152 and #33129

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by robertapplin** [comment]
Please could you send me the data used in the issue @RichardWaiteSTFC ?

**Comment by robertapplin** [comment]
The code changes look good, I just need to do some functional testing and then I'll approve this

**Comment by RichardWaiteSTFC** [comment]
@robertapplin - the data in the issue #33152 are just simulated, you only need to run the script to test (script even makes the plot).

**Comment by robertapplin** [comment]
> @robertapplin - the data in the issue #33152 are just simulated, you only need to run the script to test (script even makes the plot).

Ah yes sorry, I didn't realise the script was saving the data before loading it back in again

**Review by robertapplin — DISMISSED** [review]
The code changes look good, and the bug is fixed 👍

File: Framework/Nexus/src/NexusFileIO.cpp [hunk 1]
```diff
@@ -420,7 +420,7 @@ int NexusFileIO::writeNexusProcessedData2D(const API::MatrixWorkspace_const_sptr
     start[0] = 0;
     asize[1] = dims_array[1];
     for (size_t i = 0; i < nSpect; i++) {
-      NXputslab(fileID, localworkspace->x(i).rawData().data(), start, asize);
+      NXputslab(fileID, localworkspace->x(spec[i]).rawData().data(), start, asize);
```
Comment (martyngigg) [inline Framework/Nexus/src/NexusFileIO.cpp:423]: Can I be slightly annoying and ask that we rename this `spec` variable? I was looking over the code and I got confused whether this contains spectrum numbers or workspace indices. I checked `SaveNexusProcessed` and it passes through [`indices`](https://github.com/mantidproject/mantid/blob/d529618f2eed7c53a7c1b90bd0afb349437c514e/Framework/DataHandling/src/SaveNexusProcessed.cpp#L305). I think renaming this variable would help the readability of the code a lot and we may as well do it while while we are changing it.

**Review by martyngigg — CHANGES_REQUESTED** [review]

**Review by martyngigg — APPROVED** [review]
Thanks for making that change. I think it helps reading and understanding that code

**Pull request merged**